---
title: "Alignment and filtering"
author: "Mary Piper, Radhika Khetani, Meeta Mistry, Jihe Liu"
date: "Aug 10th, 2021"
---

Contributors: Mary Piper, Radhika Khetani, Meeta Mistry, Jihe Liu

Approximate time:

**Link to issue describing the modifications to be made:** https://github.com/hbctraining/Intro-to-ChIPseq-flipped/issues/8

## Learning Objectives

* Understand the purpose of filtering alignment reads
* Learn to perform filtering with sambamba and samtools

## Filtering reads

A key issue with a ChIP-seq data is the inclusion of multi-mapped reads (reads that are mapped to multiple loci on the reference genome). **Allowing for multiple mapped reads increases the number of usable reads and the sensitivity of peak detection; however, the number of false positives may also increase** [[1]](https://www.ncbi.nlm.nih.gov/pubmed/21779159/). Therefore, we need to filter our alignment files to **contain only uniquely mapping reads**, to increase peak-calling confidence and improve data reproducibility. As there is no parameter in Bowtie2 to keep only uniquely mapping reads, we need to perform the following steps to generate alignment files containing only the uniquely mapping reads:

1. Sort BAM files by genomic coordinates
2. Filter the reads to keep only uniquely mapping reads (this will also remove any unmapped reads)

### 2. Sort BAM files by genomic coordinates

Before we can do the filtering, we need to sort our BAM alignment files by genomic coordinates (instead of by name). To perform the sorting, we will use [Samtools](http://www.htslib.org/), a popular tool to manipulate BAM and SAM files.

The command we use is `samtools sort` with the following parameter:

* `-o`: /path/to/output/file

```bash
$ samtools sort ~/chipseq_workshop/results/bowtie2/wt_sample2_chip.bam -o ~/chipseq_workshop/results/bowtie2/wt_sample2_chip_sorted.bam \
```

### 3. Filter the reads to keep only uniquely mapping reads

Next, we can filter the sorted BAM files to keep only uniquely mapping reads. We use the `sambamba view` command with the following parameters:

* `-t`: number of threads(cores)
* `-h`: print SAM header before reads
* `-f`: format of output file (default is SAM)
* `-F`: set [custom filter](https://github.com/lomereiter/sambamba/wiki/%5Bsambamba-view%5D-Filter-expression-syntax) - we will be using the filter to remove duplicates, multimappers and unmapped reads.

```bash
$ sambamba view -h -t 2 -f bam \
-F "[XS] == null and not unmapped and not duplicate" \
~/chipseq_workshop/results/bowtie2/wt_sample2_chip_sorted.bam > ~/chipseq_workshop/results/bowtie2/wt_sample2_chip_final.bam
```

We filter out unmapped reads by specifying in the filter `not unmapped`, and duplicates with `not duplicate`. Also, among the reads that are aligned, we filter out multimappers by specifying `[XS] == null`. 'XS' is a tag generated by Bowtie2 that gives an alignment score for the second-best alignment, and it is only present if the read is aligned and more than one alignment is found for the read.

Now that the alignment files contain only uniquely mapping reads, we are ready to perform peak calling.

> _**NOTE:** After performing read alignment, it's useful to generate QC metrics for the alignment using tools such as [MultiQC](http://multiqc.info), prior to moving on to the next steps of the analysis._

> ### Filtering out Blacklisted Regions
> It is feasible to filter out the blacklisted region from our BAM files --  that is, we remove alignments that occur with defined blacklisted regions. The below code shows how to do that with `bedtools`. However, we will filter out the blacklisted region after peak calling, which serves the same function.
> ``` 
> # DO NOT RUN. DEMO CODE
> $ bedtools intersect -v -abam wt_sample2_chip_final.bam -b mm10-blacklist.v2.bed > wt_sample2_chip_final_blacklist_filtered.bam
> ```
> 
> Blacklisted regions represent artifact regions that tend to show artificially high signal (excessive unstructured anomalous reads mapping). These regions are often found at specific types of repeats such as centromeres, telomeres and satellite repeats and typically appear uniquely mappable so simple mappability filters applied above do not remove them. The ENCODE and modENCODE consortia have compiled blacklists for various species and genome versions including human, mouse, worm and fly. These blacklisted regions (coordinate files) can be filtered out from our alignment files before proceeding to peak calling.
> 
> We will revisit this in more detail when we [discuss QC metrics](https://hbctraining.github.io/Intro-to-ChIPseq/lessons/07_QC_quality_metrics.html) in a later lesson.

***
*This lesson has been developed by members of the teaching team at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.*

