---
title: "CUT&RUN data analysis"
author: "Meeta Mistry"
date: "April 19th,  2022"
---


# CUT&RUN

## A simple method for chromatin profiling

**C**leavage **U**nder **T**argets and **R**lease **U**sing **N**uclease (CUT&RUN) is an innovative chromatin mapping strategy that is rapidly gaining traction in the field. The protocol requires less than a day from cells to DNA, and can be done entirely on the benchtop using standard equipment that is already present in most molecular biology laboratories.

### How does it work?

 
1. Cells/nuclei are bound to concanavalin A–coated magnetic beads.
	- Original Henikoff paper in eLife, isolates nuclei. Using purified nuceli allows for maximized binding of antibodies to nuclear factors and will result in cleaner CUT&RUN signal compared to protocol using whole cells. 
	- More recent bioArxiv paper, whole cells are harvested. Introduce the usage of a strong detergent to permeabilize cells rather than to rely on nuclei extraction.
	
2. Cell membranes (or nuclear membrances) are permeabilized with digitonin to allow the antibody to come in and bind its target (1h to overnight)
3. The Protein A fused MNase is then added. Protein A binds the Immunoglobulin G (IgG) on the primary antibody (or mock IgG) thus targeting MNase to antibody bound proteins.
4. The nuclease is briefly activated to digest the DNA around the target protein. This targeted digestion is controlled by the addition of calcium, which MNase requires for its nuclease activity and is chelated from the reaction up until this point. The nuclease reaction is performed on ice, and only for a short period of time, thus precisely controlling the amount of cutting and mitigating noise generated by off target digestion. 
5. At this point mononucleosomal-sized DNA fragments from external organism is added (spike-in DNA).
6. Fragments are released from nuclei by a short incubation at 37 °C. 
7. These short DNA fragments can then be purified for subsequent library preparation and high-throughput sequencing.


<p align="center">
<img src="img/CR_protocol_schematic.png" width="400">
</p>

_Image source: ["AddGene Blog"](https://blog.addgene.org/cutrun-a-improved-method-for-studying-protein-dna-interactions)_

> #### What about CUT&TAG?
> For the **C**leavage **U**nder **T**argets and **Tag**mentation assay, pAG is fused to a hyperactive Tn5 transposase (pAG-Tn5) pre-loaded with sequencing adaptors, and is activated by magnesium to simultaneously fragment and “tag” antibody-labelled chromatin with adaptors. This bypasses traditional library prep steps and accelerates sample processing. However, it only works well with nuclei.
> 
> Both assays were developed in the laboratories of Dr. Steven Henikoff (Fred Hutchinson Cancer Research Center, Seattle, WA, USA) and Dr. Ulrich Laemmli (University of Geneva, Switzerland).

## CUT&RUN versus ChIP-seq

ChIP-seq is a notoriously challenging approach, in which target fragments are enriched from excess pools of sheared chromatin. Despite rigorous optimization and washing, the method is subject to high background. The resulting low signal to noise ratio makes it difficult to identify true binding sites.


### Advantages of CUT&RUN compared to ChIP-seq:
- **Requires less starting material (smaller number of cells)**. For example, while the protocol calls for 500K cells, it could be used with as low as 5K cells
- **Lower depth of sequencing**. You can produce high-quality CUT&RUN data with only 3–8 million reads per sample, compared to the 30 million or more reads required for ChIP-seq
- **Background is signifincantly reduced**, using targeted release of genomic fragments.
- **Lower costs**, by reducing antibody usage, library prep, and sequencing depth requirements


<p align="center">
<img src="img/CUT&RUN_compare.png">
</p>

_Image source: [Epicypher Blog](https://www.epicypher.com/resources/blog/cut-and-run-vs-cut-and-tag-which-one-is-right-for-you/)_

### Any limitations to CUT&RUN?

The primary limitation of CUT&RUN is the **likelihood of over-digestion of DNA** due to inappropriate timing of the Calcium-dependent MNase reaction. A similar limitation exists for contemporary ChIP-seq protocols where enzymatic or sonicated DNA shearing must be optimized.

Also, it is possible that a **chromatin complex could be too large to diffuse out** or that protein–protein interactions retain the cleaved complex. In such cases, total DNA may be extracted after the digestion.

## Controls for CUT&RUN

Typically, with ChIP-Seq it is highly recommended that each peak should be compared with the same region of the genome in a matched control sample because only a fraction of the DNA in corresponds to actual signal amidst background noise. **CUT&RUN is thought to produce a sharper stronger signal, so do we need an input control?**

The recommendation from the Henikoff papers is do not use an input control (i.e. DNA purified from cells, but without adding any antibody for enrichment). Rather they recommend the use of a nonspecific rabbit IgG antibody that will randomly coat the chromatin at low efficiency without sequence bias. With a no-antibody control, the lack of tethering increases the possibility that slight carryover of pA-MN will result in preferential fragmentation of hyperaccessible DNA.

To **assess protocol efficacy**, it is recommended to run, in parallel, a **control CUT&RUN with an antibody against a histone mark**. How does it help? After quantifying purified DNA, run it on the Bioanalyzer system. Transcription factor CUT&RUN DNA are typically around 50–150 bp and may not show up on the bioanalyzer electropherogram if run prior to amplification. This is due to the low concentration of DNA present. However, with the control histone mark CUT&RUN, should see mono-, di-, and tri-nucleosomes in the Bioanalyzer traces.



